src/Inciter/AMR/edge_store.h

```diff
@@ -0,0 +1,329 @@
+             *
+             * @param key Key of the edge to get
+             *
+             * @return A reference to the fetched edge
+             */
+            Edge_Refinement& get(edge_t key)
+            {
+                //trace_out << "get edge " << key << std::endl;
+                assert( exists(key) );
+                return edges.at(key);
+            }
+
+            Edge_Lock_Case lock_case(edge_t key)
+            {
+                return get(key).lockCase;
+            }
+
+            void erase(edge_t key)
+            {
+                edges.erase(key);
+            }
+
+            /**
+             * @brief Function to add edge to ede store
+             *
+             * @param key The key for the given edge
+             * @param e The edge data
+             *
+             * Note: This tolerate the addition of duplicate edges
+             */
+            void add(edge_t key, Edge_Refinement e)
+            {
+                // Add edge if it doesn't exist (default behavior of insert)
+                edges.insert( std::pair<edge_t, Edge_Refinement>(key, e));
+
+                // TODO: It may be worth adding a check here to ensure if we're
+                // trying to add a new edge that exists it should contain the
+                // same data
+            }
+
+            static edge_t nodes_to_key(size_t A, size_t B)
+            {
+                return edge_t(A,B);
+            }
+
+            /**
+             * @brief Function to build a  string key from two node ids
+             * NOTE: Regardless of order of arguments, the same key will be generated
+             */
+            //static std::string nodes_to_key(size_t A, size_t B)
+            //{
+                //return std::to_string(std::min(A,B)) + KEY_DELIM + std::to_string(std::max(A,B));
+            //}
+
+            /**
+             * @brief function to take the nodes representing a face
+             * and to build the possible edges based on that
+             *
+             * For a given face {ABC}, generate the edge pairs {AB, AC, BC}
+             *
+             * @param face_ids The ids of the face to generate this for
+             *
+             * @return A (partially filled) list of all edges present on the
+             * face
+             */
+            // FIXME: Is it OK that it leaves some of the array blank?
+            static edge_list_t generate_keys_from_face_ids(face_ids_t face_ids)
+            {
+                edge_list_t key_list;
+                size_t A = face_ids[0];
+                size_t B = face_ids[1];
+                size_t C = face_ids[2];
+
+                edge_t key = nodes_to_key(A,B);
+                key_list[0] = key; // TODO: Is it OK to use copy assignment here?
+
+                key = nodes_to_key(A,C);
+                key_list[1] = key;
+
+                key = nodes_to_key(B,C);
+                key_list[2] = key;
+
+                return key_list;
+            }
```

Comment:
I wonder if not generating `A->B`, `B->C`, `C->A` instead will generate the orientation which will result in negative Jacobians.